### PR TITLE
Preliminary api client and parsl apps to export data and start training task

### PR DESCRIPTION
**Pull request by rlskoeser**

### CODE REVIEW

File: src/htr2hpc/api_client.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    def export_file_url(
+        self,
+        user_id: int,
+        document_id: int,
+        document_name: str,
+        file_format: str,
+        creation_time: datetime.datetime,
+    ) -> str:
+        """Generate URL for an export file generated by :meth:`document_export`"""
+        # is there any way get current user id from api based purely on token?
```
Comment (cmroughan) [inline src/htr2hpc/api_client.py:238]: > `# is there any way get current user id from api based purely on token?`

Yes, the following API URL returns the current user only: https://test-htr.lib.princeton.edu/api/users/current/

Comment (rlskoeser) [reply src/htr2hpc/api_client.py:238]: Ah, thank you! How did you know that? Is there other documentation / code I should be looking at to understand the API better?

Comment (cmroughan) [reply src/htr2hpc/api_client.py:238]: Yes, I was viewing https://test-htr.lib.princeton.edu/api/users/ in the browser (mostly to check whether I was right to think that regular users would only ever see themselves, but admins would see everyone) when I noticed `Current` under the Extra Actions. So unfortunately no extra documentation I can point you to, but there are some helpful methods we can keep an eye out for!

Comment (rlskoeser) [reply src/htr2hpc/api_client.py:238]: Ooh! I hadn't noticed that extra actions yet! That is good to know about.

I see now that there's a documents tasks api endpoint, maybe it would be better to use that one for getting the export status...

File: src/htr2hpc/api_client.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        """paginated list of tasks"""
+        api_url = "tasks/"
+        resp = self._make_request(api_url)
+        return ResultsList(result_type="task", **resp.json())
+
+    def task(self, task_id: int):
+        """details for a single task"""
+        api_url = f"tasks/{task_id}/"
+        resp = self._make_request(api_url)
+        return to_namedtuple("task", resp.json())
```
Comment (cmroughan) [inline src/htr2hpc/api_client.py:329]: We will also want functions to locate and download model files, if a user desires to train on top of a given model. For transcription, a user can either train on top of a given transcription model or from scratch. For segmentation, a user can either train on top of a given segmentation model or "from scratch", EXCEPT here "from scratch" actually means "on top of the base blla.mlmodel that comes as the default segmenter in kraken". So we need both a way of downloading a selected model and a way of passing the [blla.mlmodel](https://github.com/mittagessen/kraken/blob/main/kraken/blla.mlmodel) along with our data.

Fortunately, given a MODEL_PK, we can simply use BASE_URL/api/models/MODEL_PK/ and then get the URL to download the model in the "file" field. `requests.get()` should then work just fine.

For blla.mlmodel, we can just grab it from the kraken install, or store a copy somewhere.

File: src/htr2hpc/train.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        "base_url",
+        help="Base URL for eScriptorium instance (do not include /api/)",
+        type=str,
+    )
+    parser.add_argument(
+        "-d", "--document", help="Document id to export", type=int, dest="document_id"
+    )
+    parser.add_argument(
+        "-t", "--transcription", help="Transcription id to export", type=int
+    )
```
Comment (cmroughan) [inline src/htr2hpc/train.py:38]: A few suggested additions:
```
parser.add_argument(
	"-j", "--job", help="Job type (segmentation or transcription)", type=str
)
parser.add_argument(
	"-m", "--model", help="Optional model id to export (if fine-tuning)", type=int
)
parser.add_argument(
	"-p", "--parts", help="Optional list of parts to export (if not entire document)"
)
```

File: src/htr2hpc/train.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    args = parser.parse_args()
+
+    parsl.load(config)
+    training_data = prep_training_data(
+        args.base_url,
+        api_token,
+        args.document_id,
+        args.transcription,
+    )
+
```
Comment (cmroughan) [inline src/htr2hpc/train.py:48]: Probably add logic like: if args.model exists, then run a prep_model() function taking that MODEL_PK as input and saving the model as `model_data` or something. That would handle both cases of segmentation and transcription training where the user chooses to finetune on an existing model.

To handle segmentation training "from scratch" (which is actually on top of the default blla.mlmodel), we would want logic like: if args.model does not exist and args.job == 'segmentation', then load in blla.mlmodel instead.

File: src/htr2hpc/train_apps.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    data_dir = pathlib.Path("training_data")
+    data_dir.mkdir(exist_ok=True)
+
+    # extract everything in the zipfiile to the training data dir
+    with ZipFile(export_zipfile) as zip_exp:
+        zip_exp.extractall(path=data_dir)
+
+    # return training directory as a parsl file
+    return File(data_dir)
+
```
Comment (cmroughan) [inline src/htr2hpc/train_apps.py:33]: Should probably also have a `prep_model()` function, taking as input `es_base_url`, `es_api_token`, `model_id`.

File: src/htr2hpc/train_apps.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+@bash_app
+def segtrain(inputs=[], outputs=[]):
```
Comment (cmroughan) [inline src/htr2hpc/train_apps.py:36]: We could put both transcription and segmentation training in one function, but we probably want a separate `train` function for transcription training. I'm laying out the usual eScriptorium train/segtrain commands as they would be formatted in command-line kraken below.

- A command for transcription training from scratch:

  - `ketos train --epochs 50 -o model --workers HOW_MANY_CPUS -d cuda:0 -f alto {input_data_dir}/*.xml`
  - (For workers, use the number of CPUs requested in the slurm job.)
- A command for transcription refining on another model:
  - `ketos train --epochs 50 --resize union -i {path_to_starting_model} -o model --workers HOW_MANY_CPUS -d cuda:0 -f alto {input_data_dir}/*.xml`
- A command for segmentation training "from scratch":
  - `ketos segtrain --epochs 50 --resize both -i {path_to_blla_mlmodel} -o model --workers HOW_MANY_CPUS -d cuda:0 -f xml {input_data_dir}/*.xml`
- A command for segmentation refining on another model:
  - `ketos segtrain --epochs 50 --resize both -i {path_to_starting_model} -o model --workers HOW_MANY_CPUS -d cuda:0 -f xml {input_data_dir}/*.xml`

File: src/htr2hpc/api_client.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        api_url = "users/current/"
+        return to_namedtuple("user", self._make_request(api_url).json())
+
+    def models(self):
+        """paginated list of models"""
+        api_url = "models/"
+        resp = self._make_request(api_url)
+        return ResultsList(result_type="model", **resp.json())
+
+    def update_model(self, model_id: int, model_file: pathlib.Path):
```
[the remark this answers is not in the document]
Comment (cmroughan) [reply src/htr2hpc/api_client.py:191]: I think for POSTing a new model we'd want something like this, they're easy enough changes to make. The problem just remains that I can't get accuracy_percent to update. (The upload goes through, it just means there's no recorded accuracy in the model list which is... not ideal.)

```
    def upload_model(self, model_file: pathlib.Path, model_name: str, model_job: str):
        """Upload the new model file to a user's list of models."""
        api_url = f"models/"
        with open(model_file, "rb") as mfile:
            files = {"file": mfile}
            data = {
                "name": model_name, # required, get from user
                "job": model_job,  # required, either "Segment" or "Recognize", get from user
                "accuracy_percent": 0.0, # 0.0 by default; can be found in the mlmodel file
                # file_size (int); not required when POSTing with API, will be automatically filled in
                # versions ? does not seem to be used
            }
            resp = self._make_request(api_url, method="POST", files=files, data=data)
        # on successful update, returns the model object
        return to_namedtuple("model", resp.json())
```

File: src/htr2hpc/train.py (hunk 3 above)
Comment (rlskoeser) [reply src/htr2hpc/train.py:38]: @cmroughan these look like good options to add, do you want to use the suggest changes so it will show as your commit? (and/or you can send me your co-author commit email)

Comment (cmroughan) [reply src/htr2hpc/train.py:38]: Ah yes, I did forget about the add a suggestion option!

File: src/htr2hpc/api_client.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            # (useful for dev if nothing else...)
+            logger.error(resp.status_code)
+            logger.error(resp.content)
+
+    def current_user(self):
+        """Get information about the current user account"""
+        api_url = "users/current/"
+        return to_namedtuple("user", self._make_request(api_url).json())
+
+    def models(self):
```
[the remark this answers is not in the document]
Comment (laurejt) [reply src/htr2hpc/api_client.py:185]: Yes, that sounds like a good idea. Giving methods names that are more descriptive of their function should hopefully make it easier to use.

File: src/htr2hpc/api_client.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
+import datetime
+import logging
+from collections import namedtuple
+from dataclasses import dataclass
+from time import sleep
+import pathlib
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:6]: Perhaps only import `Path`
```suggestion
from pathlib import Path
```

File: src/htr2hpc/api_client.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+        # otherwise, create it and add it to the registry
+        if nt_class is None:
+            nt_class = namedtuple(name, data)
+            RESULTCLASS_REGISTRY[name] = nt_class
+        # once we have the class, initialize an instance with the given data
+        return nt_class(
+            # convert any nested objects to namedtuple classes
+            # use key name as namedtuple class name; convert plurals to singular
+            # NOTE: could add aliases other cleanup, e.g. shared_with_user(s) == user
+            **{key: to_namedtuple(key.rstrip("s"), val) for key, val in data.items()}
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:98]: Are there any expected names where stripping the ending "s" would be incorrect?

File: src/htr2hpc/api_client.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+                setattr(self, date_field, datetime.datetime.fromisoformat(value))
+
+    def duration(self) -> datetime.timedelta | None:
+        "how long the task took to complete"
+        if self.done_at is not None:
+            return self.done_at - self.started_at
+
+
+# keep a registry of result classes for API result objects,
+# so they can be reused once they are defined
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:75]: How varied are these result classes? Do these change enough that it's not possible to define them upfront?

File: src/htr2hpc/api_client.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+        method: str = "GET",
+    ):
+        """
+        Make a GET request with the configured session. Takes a url
+        relative to :attr:`api_root` and optional dictionary of parameters for the request.
+        """
+        rqst_url = f"{self.api_root}/{url}"
+        rqst_opts = {}
+        if params:
+            rqst_opts["params"] = params.copy()
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:141]: Is this a flat dict? If not, consider `deepcopy`

File: src/htr2hpc/api_client.py [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+                rqst_opts["files"] = files
+        else:
+            raise ValueError(f"unsupported http method: {method}")
+
+        resp = session_request(rqst_url, **rqst_opts)
+        logger.debug(
+            "get %s %s: %f sec",
+            rqst_url,
+            resp.status_code,
+            resp.elapsed.total_seconds(),
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:160]: Perhaps use f-string
```suggestion
            f"get {rqst_url} {resp.status_code}: {resp.elapsed.total_seconds()} sec"
```

File: src/htr2hpc/api_client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (laurejt) [reply src/htr2hpc/api_client.py:185]: Maybe `get_model_list`?

File: src/htr2hpc/api_client.py [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+        # is a based on block type ids for this document
+        # and may also include "Undefined" and "Orphan"
+        document_info = self.document(document_id)
+        block_types = [block.pk for block in document_info.valid_block_types]
+        block_types.extend(["Undefined", "Orphan"])
+
+        data = {
+            "transcription": transcription_id,  # could be multiple(?)
+            "file_format": "alto",
+            "include_images": True,
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:234]: Should this use the input arg `include_images`?
```suggestion
            "include_images": include_images,
```

File: src/htr2hpc/api_client.py [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+
+        # ... get document name from api by document id?
+
+        # NOTE: filename logic copied & adapted from escriptorium
+        # import.export.BaseExporter logic
+        # NOTE2: escriptorium code uses datetime.now() so there's no guarantee
+        # this will match the completion time of the task...
+        base_filename = "export_doc%d_%s_%s_%s" % (
+            document_id,
+            slugify(document_name).replace("-", "_")[:32],
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:261]: Can you explain this in more detail? Using `slugify` seems like overkill, or at the very least confirming which characters are being removed and/or replaced.

File: src/htr2hpc/api_client.py [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+        # is there any way get current user id from api based purely on token?
+        # if not, consider using users api endpoint to get numeric id from username
+
+        # ... get document name from api by document id?
+
+        # NOTE: filename logic copied & adapted from escriptorium
+        # import.export.BaseExporter logic
+        # NOTE2: escriptorium code uses datetime.now() so there's no guarantee
+        # this will match the completion time of the task...
+        base_filename = "export_doc%d_%s_%s_%s" % (
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:259]: Consider replacing with an f-string

File: src/htr2hpc/api_client.py [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+            document_id=document_id,
+            transcription_id=transcription_ids,
+            include_images=True,
+        )
+        # currently returns status=ok, indicating a task has been queued
+        if result.status == "ok":
+            task_list = self.tasks()
+            # assume most recent task, which is listed first
+            # TODO: confirm correct task id based on method (export) and document name
+            export_task = task_list.results[0]
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:287]: Do you only need the first item from the results list?

File: src/htr2hpc/api_client.py [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+    document: int
+    document_part: int
+    workflow_state: int
+    label: str
+    messages: str
+    queued_at: datetime.datetime
+    started_at: datetime.datetime
+    done_at: datetime.datetime
+    method: str
+    user: int
```
Comment (laurejt) [inline src/htr2hpc/api_client.py:50]: Perhaps worth some comments, especially for the cases where these are identifiers rather than contents or filenames.

File: src/htr2hpc/train.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    )
+    parser.add_argument(
+	"-j", "--job", help="Job type (segmentation or transcription)", type=str
+    )
+    parser.add_argument(
+	"-m", "--model", help="Optional model id to export (if fine-tuning)", type=int
+    )
+    parser.add_argument(
+	"-p", "--parts", help="Optional list of parts to export (if not entire document)"
+    )
```
Comment (laurejt) [inline src/htr2hpc/train.py:47]: I'm guessing this component has been worked out yet, but I'm guessing this may use `nargs=+`

File: src/htr2hpc/api_client.py (hunk 18 above)
Comment (laurejt) [reply src/htr2hpc/api_client.py:50]: It seems like it might be worth renaming `document` to `document_id`
